### PR TITLE
BUG: aggregations were getting overwritten if they had the same name

### LINKED ISSUE #30092: Can't use named aggregation with resample.
#### Code Sample, a copy-pastable example if possible
```python
import pandas as pd
import numpy as np
from functools import partial

df = pd.DataFrame(
    np.random.randn(1000, 3), 
    index=pd.date_range('1/1/2012', freq='S', periods=1000), 
    columns=['A', 'B', 'C']
)
dfg = df.resample('3T').agg(
    {'A': [
        partial(np.quantile, q=.9999), 
        partial(np.quantile, q=.90),
    ]}
)
```
#### Problem description
Resample, unlike groupby, has no ability to do named aggregation. This means if you use a function twice on the same column you get an error saying:

``pandas.core.base.SpecificationError: Function names must be unique, found multiple named quantile``. 

It would be very useful to make the resample agg interface similar to the groupby agg interface. An example of the named aggregation is here: https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html#named-aggregation

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.9.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : en_US.UTF-8
pandas           : 0.25.3
numpy            : 1.17.4
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 18.1
setuptools       : 40.6.2
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 0.15.1
pytables         : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
</details>

### LINKED ISSUE #30880: Multiple aggregations with the same name get overwritten
#### Code Sample, a copy-pastable example if possible

```
import pandas as pd
import numpy as np
from functools import partial

df = pd.DataFrame(
    np.random.randn(1000, 3), 
    index=pd.date_range('1/1/2012', freq='S', periods=1000), 
    columns=['A', 'B', 'C']
)
dfg = df.resample('3T').agg(
    {'A': [
        partial(np.quantile, q=.9999), 
        partial(np.quantile, q=.1111),
    ]}
)
```
returns
```
                            A          
                     quantile  quantile
2012-01-01 00:00:00 -1.035997 -1.035997
2012-01-01 00:03:00 -1.326052 -1.326052
2012-01-01 00:06:00 -1.267661 -1.267661
2012-01-01 00:09:00 -1.321961 -1.321961
2012-01-01 00:12:00 -1.103027 -1.103027
2012-01-01 00:15:00 -0.987614 -0.987614
```

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit           : 5e9fe4e0843266cb1c64bff4af70b3665a4d6a8d
python           : 3.7.3.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-74-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 0.25.0+1719.g5e9fe4e08.dirty
numpy            : 1.17.3
pytz             : 2019.3
dateutil         : 2.8.0
pip              : 19.3.1
setuptools       : 42.0.2.post20191201
Cython           : 0.29.14
pytest           : 5.3.2
hypothesis       : 4.55.4
sphinx           : 2.3.0
blosc            : None
feather          : None
xlsxwriter       : 1.2.6
lxml.etree       : 4.4.2
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.9.0
pandas_datareader: None
bs4              : 4.8.1
bottleneck       : 1.3.1
fastparquet      : 0.3.2
gcsfs            : None
lxml.etree       : 4.4.2
matplotlib       : 3.1.2
numexpr          : 2.7.0
odfpy            : None
openpyxl         : 3.0.1
pandas_gbq       : None
pyarrow          : 0.15.1
pytables         : None
pytest           : 5.3.2
s3fs             : 0.4.0
scipy            : 1.4.0
sqlalchemy       : 1.3.11
tables           : 3.6.1
tabulate         : 0.8.6
xarray           : 0.14.1
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.6
numba            : 0.46.0
</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [x] closes #30880 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

xref #30092

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -318,13 +319,18 @@ def _aggregate_multiple_funcs(self, arg):
...
-        if any(isinstance(x, DataFrame) for x in results.values()):
+        if any(isinstance(x.aggregation, DataFrame) for x in results):
             # let higher level handle
-            return results
+            return dict(results)
 
-        return DataFrame(results, columns=columns)
+        # If there are multiple aggregations with the same name, we need to pass
+        # them to the constructor with different keys.
+        df = DataFrame({n: x.aggregation for n, x in enumerate(results)}).rename(
```
Comment (WillAyd) [inline pandas/core/groupby/generic.py]: We have similar functionality already built in to do this. If you keep `results` as a dict and use a `base.OutputKey` with the results as a value you can ultimately just call `self._wrap_aggregated_output`

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by charlesdong1991** [comment]
nice, thanks for the PR @MarcoGorelli 

I think this PR deserves a new issue other than #30092 , so would suggest to `xref` it.

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py]: @WillAyd nice, thanks!

I've updated the PR with that instead, is there a better way of writing this line:
```
{key.label: value for key, value in results.items()}
```
?

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -318,13 +318,15 @@ def _aggregate_multiple_funcs(self, arg):
...
-            results[name] = obj.aggregate(func)
+            results[base.OutputKey(label=name, position=idx)] = obj.aggregate(func)
 
         if any(isinstance(x, DataFrame) for x in results.values()):
             # let higher level handle
-            return results
+            return {key.label: value for key, value in results.items()}
 
-        return DataFrame(results, columns=columns)
+        if results:
```
Comment (WillAyd) [inline pandas/core/groupby/generic.py]: Is _wrap_aggregated_output failing in results is empty? If so I think that should be fixed rather than branching here

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 3]
```diff
@@ -239,6 +239,56 @@ def test_agg_multiple_functions_maintain_order(df):
     tm.assert_index_equal(result.columns, exp_cols)
 
 
+def test_agg_multiple_functions_same_name(df):
```
Comment (WillAyd) [inline pandas/tests/groupby/aggregate/test_aggregate.py]: Just a general comment on test; if you can make shorter helps readability (i.e. 3 rows of data for test vs 15 can only help)

**Review by WillAyd — CHANGES_REQUESTED** [review]
Nice job looks pretty good

File: pandas/core/groupby/generic.py [hunk 4]
```diff
@@ -318,13 +318,15 @@ def _aggregate_multiple_funcs(self, arg):
                 obj = copy.copy(obj)
                 obj._reset_cache()
                 obj._selection = name
-            results[name] = obj.aggregate(func)
+            results[base.OutputKey(label=name, position=idx)] = obj.aggregate(func)
```
Comment (jreback) [inline pandas/core/groupby/generic.py]: you can type results when its created

File: doc/source/whatsnew/v1.0.0.rst [hunk 5]
```diff
@@ -1112,6 +1112,7 @@ Groupby/resample/rolling
 - Bug in :meth:`DataFrame.groupby` when using nunique on axis=1 (:issue:`30253`)
 - Bug in :meth:`GroupBy.quantile` with multiple list-like q value and integer column names (:issue:`30289`)
 - Bug in :meth:`GroupBy.pct_change` and :meth:`core.groupby.SeriesGroupBy.pct_change` causes ``TypeError`` when ``fill_method`` is ``None`` (:issue:`30463`)
+- Bug in :meth:`SeriesGroupBy._aggregate_multiple_funcs` was resulting in aggregations being overwritten when they shared the same name (:issue:`30092`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: this should be a user facing note

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by MarcoGorelli** [comment]
@jreback @WillAyd thanks for your reviews, have updated accordingly

File: pandas/core/groupby/generic.py [hunk 6]
```diff
@@ -358,8 +357,10 @@ def _wrap_series_output(
         if len(output) > 1:
             result = DataFrame(indexed_output, index=index)
             result.columns = columns
-        else:
+        elif not columns.empty:
             result = Series(indexed_output[0], index=index, name=columns[0])
+        else:
```
Comment (WillAyd) [inline pandas/core/groupby/generic.py]: Do we need this branch? Is there a test that hits this?

File: doc/source/whatsnew/v1.1.0.rst [hunk 7]
```diff
@@ -119,7 +119,7 @@ Numeric
 Conversion
 ^^^^^^^^^^
 - Bug in :class:`Series` construction from NumPy array with big-endian ``datetime64`` dtype (:issue:`29684`)
--
+- Bug in :meth:`DataFrame.replace` was changing other columns' dtypes when values in one column were being replaced with ``NaN`` (:issue:`30512`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.1.0.rst]: need to remove this

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 6 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py]: @WillAyd Here's a test that hits it:
```
pytest pandas/tests/groupby/aggregate/test_aggregate.py::TestNamedAggregationSeries::test_no_args_raises
```

**Comment by MarcoGorelli** [comment]
Hi @WillAyd - sorry to chase you up, just wanted to ask if there's anything else that needs doing here or if it's alright (or indeed if it's the wrong fix altogether :) )

File: pandas/core/groupby/generic.py [hunk 8]
```diff
@@ -311,8 +311,8 @@ def _aggregate_multiple_funcs(self, arg):
 
             arg = zip(columns, arg)
 
-        results = {}
-        for name, func in arg:
+        results: Mapping[base.OutputKey, Union[Series, DataFrame]] = {}
```
Comment (WillAyd) [inline pandas/core/groupby/generic.py]:
```suggestion
        results: Dict[base.OutputKey, Union[Series, DataFrame]] = {}
```

Nice idea but only need mapping for API exposed signatures; internally if its a dict can just type as such

File: pandas/core/groupby/generic.py (hunk 6 above)
Comment (WillAyd) [reply pandas/core/groupby/generic.py]: I was referring to the `result = DataFrame()` line - seems very strange to return this from this function

File: pandas/core/groupby/generic.py [hunk 9]
```diff
@@ -321,13 +321,12 @@ def _aggregate_multiple_funcs(self, arg):
...
-            results[name] = obj.aggregate(func)
+            results[base.OutputKey(label=name, position=idx)] = obj.aggregate(func)
 
         if any(isinstance(x, DataFrame) for x in results.values()):
             # let higher level handle
-            return results
-
-        return DataFrame(results, columns=columns)
+            return {key.label: value for key, value in results.items()}
+        return DataFrame(self._wrap_aggregated_output(results), columns=columns)
```
Comment (WillAyd) [inline pandas/core/groupby/generic.py]: Is the DataFrame constructor still required here?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 9 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py]: In the test
```
pytest pandas/tests/groupby/aggregate/test_aggregate.py::test_aggregate_item_by_item
```
when we get here we have
```
(Pdb) results
{OutputKey(label='<lambda>', position=0): A
bar    3
foo    5
Name: B, dtype: int64}
(Pdb) self._wrap_aggregated_output(results)
A
bar    3
foo    5
Name: <lambda>, dtype: int64
(Pdb) type(self._wrap_aggregated_output(results))
<class 'pandas.core.series.Series'>
```

File: pandas/core/groupby/generic.py (hunk 6 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py]: In the test
```
pytest pandas/tests/groupby/aggregate/test_aggregate.py::TestNamedAggregationSeries::test_no_args_raises
```
we reach this branch. Before, it would hit the "else" branch, and would raise an error because `columns[0]` would be out of range - hence, I added this extra branch.

But we can avoid ever getting here by doing a check on `results` in `_aggregate_multiple_funcs`, so I'll do that instead

File: pandas/core/groupby/generic.py (hunk 9 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py]: @WillAyd have updated with a call to `.to_frame` (if necessary)

File: pandas/core/groupby/generic.py [hunk 10]
```diff
@@ -321,13 +321,18 @@ def _aggregate_multiple_funcs(self, arg):
...
         if any(isinstance(x, DataFrame) for x in results.values()):
             # let higher level handle
-            return results
+            return {key.label: value for key, value in results.items()}
 
-        return DataFrame(results, columns=columns)
+        if not results:
+            return DataFrame()
+        output = self._wrap_aggregated_output(results)
+        if not isinstance(output, ABCDataFrame):
```
Comment (jreback) [inline pandas/core/groupby/generic.py]: hmm, what case hits this?

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 11]
```diff
@@ -250,6 +250,24 @@ def test_agg_multiple_functions_maintain_order(df):
...
+    # GH 30880
+    np.random.seed(1)
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py]: why is this needed?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 11 above)
Comment (MarcoGorelli) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: No longer needed, sorry, came from a previous commit

File: pandas/core/groupby/generic.py (hunk 10 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py]: @jreback here's one:

pandas/tests/groupby/test_categorical.py::test_groupby_agg_categorical_columns

```
    func = pd.Series.nunique
    df = pd.DataFrame(
        {
            "id": [0, 1, 2, 3, 4],
            "groups": [0, 1, 1, 2, 2],
            "value": pd.Categorical([0, 0, 0, 0, 1]),
        }
    ).set_index("id")
    result = df.groupby("groups").agg(func)
```

File: pandas/core/groupby/generic.py [hunk 12]
```diff
@@ -321,13 +321,18 @@ def _aggregate_multiple_funcs(self, arg):
                 obj = copy.copy(obj)
                 obj._reset_cache()
                 obj._selection = name
-            results[name] = obj.aggregate(func)
+            results[base.OutputKey(label=name, position=idx)] = obj.aggregate(func)
 
         if any(isinstance(x, DataFrame) for x in results.values()):
             # let higher level handle
-            return results
+            return {key.label: value for key, value in results.items()}
```
Comment (WillAyd) [inline pandas/core/groupby/generic.py]: Do you know the code path that hits this? Wouldn't this run into the same error if `key.label` is non-unique?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 12 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py]: This happens if one of the aggregations is 'ohlc' - will look into this, thanks

**Comment by MarcoGorelli** [comment]
@WillAyd @jreback  on a second thought, is it even desirable to support this? Having identically labelled columns would bring problems later on anyway, perhaps just raising an error message if there are identically named functions would be better?

**Comment by WillAyd** [comment]
While weird we do technically allow for duplicate column labels, so I think OK to move forward. Looks like a few outstanding items, so let us know if you get tripped up while investigating

**Comment by MarcoGorelli** [comment]
@WillAyd Here's something that's tripping me up: if I have
```
>>> import functools
>>> import pandas as pd
>>> import pandas._testing as tm
>>> import numpy as np

>>> df = tm.makeTimeDataFrame().head()
>>> df
                   A         B         C         D
2000-01-03 -1.629752 -1.277942 -0.864068 -0.693202
2000-01-04  0.661794 -0.708207  1.399000 -0.174563
2000-01-05  0.222601  1.286268  1.011485  0.055014
2000-01-06  0.031587  0.253470 -0.327816  1.396794
2000-01-07 -0.502560 -0.243002  1.266337  0.055613
```

Then, if I do
```

```
>>> df.resample('3D').agg(
...     {'A': ['ohlc', functools.partial(np.std, ddof=0), functools.partial(np.std, ddof=1)]}
... )
```
then, when I get to line 284 in `pandas/core/groupby/generic.py`:
```
        if isinstance(ret, dict):
            from pandas import concat

            ret = concat(ret, axis=1)
```
I have:
```
>>> ret
{OutputKey(label='ohlc', position=0):                 open      high       low     close
2000-01-03 -1.629752  0.661794 -1.629752  0.222601
2000-01-06  0.031587  0.031587 -0.502560 -0.502560, OutputKey(label='std', position=1): 2000-01-03    0.993047
2000-01-06    0.267074
Freq: 3D, Name: A, dtype: float64, OutputKey(label='std', position=2): 2000-01-03    1.216230
2000-01-06    0.377699
Freq: 3D, Name: A, dtype: float64}

>>> concat(ret, axis=1)
                ohlc                                     std          
                   0                                       1         2
                open      high       low     close         A         A
2000-01-03 -1.629752  0.661794 -1.629752  0.222601  0.993047  1.216230
2000-01-06  0.031587  0.031587 -0.502560 -0.502560  0.267074  0.377699
>>> 
```

The problem is that there is still the position number - if I try changing the code so it says
```
ret = concat(ret.values(), axis=1, keys=[key.label for key in ret.keys()])
```
then I get the error
```
*** TypeError: int() argument must be a string, a bytes-like object or a number, not 'slice'
```

All I could think of to get around it is:
```
            ret = concat(ret.values(), axis=1, keys=[key.position for key in keys])
            ret.columns = ret.columns.set_levels([key.label for key in keys], level=0, verify_integrity=False)
```
, which doesn't seem right

**Comment by WillAyd** [comment]
Does that only happen with ohlc? Agreed the position number shouldn't be there; may something off in the groupby `_wrap_*` functions

File: pandas/core/groupby/generic.py (hunk 10 above)
Comment (jreback) [reply pandas/core/groupby/generic.py:333]: hmm i think this last case should be handled in _wrap_aggregated_output, we shouldn't just force to a DataFrame, something else is going on

**Review by jreback — CHANGES_REQUESTED** [review]
can you also add the test exactly from here: https://github.com/pandas-dev/pandas/issues/30092 (and add this issue to the whatsnew)

File: pandas/core/groupby/generic.py (hunk 10 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py:333]: Aren't there cases where we would want `_wrap_aggregated_output` to return a Series though?

For the test `pandas/tests/groupby/test_categorical.py::test_groupby_agg_categorical_columns`, when we enter `_wrap_aggregated_output`, we have:
```
>>> output
{OutputKey(label='value', position=0): array([1, 1, 2])}
>>> self
<pandas.core.groupby.generic.SeriesGroupBy object at 0x7f7f7f178730>
```

while for the test `pandas/tests/groupby/aggregate/test_aggregate.py::test_agg_ser_multi_key`, when we enter it, we have
```
>>> output
{OutputKey(label='C', position=0): array([ 0.3004091 , -1.87694605,  0.7516694 ,  1.01936873,  0.7758193 ,
       -0.60044377])}
>>> self
self
<pandas.core.groupby.generic.SeriesGroupBy object at 0x7f2963500160>
```

Any suggestions on how to distinguish when we'd want to return a DataFrame and when a Series?

**Comment by MarcoGorelli** [comment]
> Does that only happen with ohlc?

@WillAyd ohlc is the only one I've come across - are there other operations that expand dimensions?

**Comment by MarcoGorelli** [comment]
> The problem is that there is still the position number in the columns - if I change the code so it says
> ```
> ret = concat(ret.values(), axis=1, keys=[key.label for key in ret.keys()])
> ```
> then I get the error
> ```
> *** TypeError: int() argument must be a string, a bytes-like object or a number, not 'slice'
> ```

The problem here is that when we go to initialise the MultiIndex for the columns, then in `_make_concat_multiindex` from `pandas/core/reshape/concat.py` we have that `levels` is
```python
>>> levels
[Index(['ohlc', 'std', 'std'], dtype='object'),
 Index(['A', 'close', 'high', 'low', 'open'], dtype='object')]
```
and so the problem comes from
```python
>>> levels[0].get_loc('std')
slice(1, 3, None)
```

**Comment by MarcoGorelli** [comment]
> can you also add the test exactly from here: #30092 (and add this issue to the whatsnew)

I'm not sure how it's different - #30092 doesn't raise a `SpecificationError` anymore in 1.0.3. I've added a test to make sure it doesn't break again, but do we need a whatsnew issue for that?

**Comment by MarcoGorelli** [comment]
> > The problem is that there is still the position number in the columns - if I change the code so it says
> > ```
> > ret = concat(ret.values(), axis=1, keys=[key.label for key in ret.keys()])
> > ```
> > 
> > 
> > then I get the error
> > ```
> > *** TypeError: int() argument must be a string, a bytes-like object or a number, not 'slice'
> > ```
> 
> The problem here is that when we go to initialise the MultiIndex for the columns, then in `_make_concat_multiindex` from `pandas/core/reshape/concat.py` we have that `levels` is
> 
> ```python
> >>> levels
> [Index(['ohlc', 'std', 'std'], dtype='object'),
>  Index(['A', 'close', 'high', 'low', 'open'], dtype='object')]
> ```
> 
> and so the problem comes from
> 
> ```python
> >>> levels[0].get_loc('std')
> slice(1, 3, None)
> ```

Have raised #33654 for this, I think I need to solve that one first

**Comment by MarcoGorelli** [comment]
@jreback @WillAyd thanks for your reviews. Now that #33654 is closed, I've been able to update this so that the "ohlc" case passes.

I haven't been able to get rid of the special-casing of the Series case, i.e. these lines:
```python
if not isinstance(output, ABCDataFrame):
    output = output.to_frame()
```

On master the following is returned
```python
DataFrame(results)
```

But in this PR, we go through `_wrap_aggregated_output`, which may return a Series if it's input is of length 1. Do you have any suggestions on how to get around this?

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 13]
```diff
@@ -252,6 +255,71 @@ def test_agg_multiple_functions_maintain_order(df):
...
+            ("A", "std", "A"),
+            ("A", "std", "A"),
+        ]
+    )
+    expected_values = np.hstack([df.resample("3D").A.ohlc(), expected_values])
+    expected = pd.DataFrame(
+        expected_values, columns=expected_columns, index=expected_index
+    )
+    with tm.assert_produces_warning(PerformanceWarning):
+        tm.assert_frame_equal(result, expected)
```
Comment (MarcoGorelli) [inline pandas/tests/groupby/aggregate/test_aggregate.py:299]: This is thrown by
```python
assert col in right
```
in `assert_frame_equal`

File: pandas/core/groupby/generic.py [hunk 14]
```diff
@@ -321,13 +321,18 @@ def _aggregate_multiple_funcs(self, arg):
...
 
         if any(isinstance(x, DataFrame) for x in results.values()):
             # let higher level handle
             return results
 
-        return DataFrame(results, columns=columns)
+        if not results:
+            return DataFrame()
+        output = self._wrap_aggregated_output(results)
+        if not isinstance(output, ABCDataFrame):
```
Comment (jreback) [inline pandas/core/groupby/generic.py:333]: you can use DataFrame here (rather than ABC)

File: pandas/core/groupby/generic.py (hunk 10 above)
Comment (jreback) [reply pandas/core/groupby/generic.py:333]: see my comment above, _wrap_aggregated_output should handle this

**Review by jreback — CHANGES_REQUESTED** [review]
can you also add the actual test from the OP

File: pandas/core/groupby/generic.py (hunk 10 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py:333]: Hi @jreback - thanks for your feedback. Sorry, I don't understand what you mean here - I can't remove
```python
if not isinstance(output, ABCDataFrame):
    output = output.to_frame()
```
from here because then `pandas/tests/groupby/test_categorical.py::test_groupby_agg_categorical_columns` would fail, as in that test `_wrap_aggregated_output` needs to return a Series.

Also, I make this change
```diff
-        return self._reindex_output(result)
+        reindexed_output = self._reindex_output(result)
+        if not isinstance(reindexed_output, DataFrame):
+            reindexed_output = reindexed_output.to_frame()
+        return reindexed_output
```
to  `_wrap_aggregated_output` because there are cases (e.g. `pandas/tests/groupby/test_categorical.py::test_groupby_agg_categorical_columns`) when `_wrap_aggregated_output` _should_ return a Series.

**Comment by MarcoGorelli** [comment]
> can you also add the actual test from the OP

Sure, done. Now the first part of the test matches #30880 exactly, and the second part just has `ohlc` added in

**Comment by MarcoGorelli** [comment]
Have merged master. With the most recent changes there's no need to special-case `Series`, so have removed that branch

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 15]
```diff
@@ -252,6 +255,48 @@ def test_agg_multiple_functions_maintain_order(df):
...
+    expected_columns = MultiIndex.from_tuples([("A", "quantile"), ("A", "quantile")])
+    expected_values = expected_values = np.array(
+        [df.resample("3T").A.quantile(q=q).values for q in [0.9999, 0.1111]]
+    ).T
+    expected = pd.DataFrame(
+        expected_values, columns=expected_columns, index=expected_index
+    )
+    tm.assert_frame_equal(result, expected)
+
+    # check what happens if ohlc (which expands dimensions) is present
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py:278]: can you make this a separate test starting here (with an appropraitea name)

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 16]
```diff
@@ -252,6 +255,48 @@ def test_agg_multiple_functions_maintain_order(df):
...
+            ("A", "ohlc", "close"),
+            ("A", "quantile", "A"),
+            ("A", "quantile", "A"),
+        ]
+    )
+    expected_values = np.hstack([df.resample("3T").A.ohlc(), expected_values])
+    expected = pd.DataFrame(
+        expected_values, columns=expected_columns, index=expected_index
+    )
+    with tm.assert_produces_warning(PerformanceWarning):
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py:296]: can you add a comment why we have a PerformanceWarning

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. testing change request. ping on green.

**Comment by MarcoGorelli** [comment]
@jreback thanks for your review - sure, done.

>  ping on green.

TravisCI is green if I click on 'details', but for some reason still shows pending here in the summary

**Comment by MarcoGorelli** [comment]
Before this is merged, I should add a test case using pd.NamedAgg, xref #34380

**Comment by MarcoGorelli** [comment]
> Before this is merged, I should add a test case using pd.NamedAgg, xref #34380

@jreback, have added a test which uses pd.NamedAgg, as in #34380, it's green now (though again, the Travis CI check still shows 'pending' even though when I click on it it says "#80225 passed "

**Comment by MarcoGorelli** [comment]
friendly ping :)

File: pandas/core/groupby/generic.py [hunk 17]
```diff
@@ -321,13 +321,16 @@ def _aggregate_multiple_funcs(self, arg):
...
-            results[name] = obj.aggregate(func)
+            results[base.OutputKey(label=name, position=idx)] = obj.aggregate(func)
 
         if any(isinstance(x, DataFrame) for x in results.values()):
             # let higher level handle
             return results
 
-        return self.obj._constructor_expanddim(results, columns=columns)
+        if not results:
+            return DataFrame()
```
Comment (jreback) [inline pandas/core/groupby/generic.py:331]: hmm is this correct? do we have tests that hit this. I would think we would have *somthing* e.g. columns even if this is empty

Comment (jreback) [reply pandas/core/groupby/generic.py:331]: also why is this not just handled in wrap_aggregated_output?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 17 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py:331]: Here's a test that hits it: `pandas/tests/groupby/aggregate/test_aggregate.py::TestNamedAggregationSeries::test_no_args_raises`

When trying to move this to `wrap_aggregated_output` I ran into #34977, so I'll try to address that first

Comment (jreback) [reply pandas/core/groupby/generic.py:331]: this still is quite fishy . if you pass en empty result to `self._wrap_aggregated_output` what do you get as `output`? I really don't like special cases like this which inevitably hide errors and make groking code way more complex.

So prefer to have _wrap_aggregated_output handle this correctly. you *may* not even need L333, its possible to pass columns to _wrap_aggregated_output

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 17 above)
Comment (jreback) [reply pandas/core/groupby/generic.py:331]: i would just fix this, need to check if len(indexed_output)

Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py:331]: So,
```python
elif len(indexed_output):
    result = self.obj._constructor(
        indexed_output[0], index=index, name=columns[0]
    )
else:
    result = self.obj._constructor()
```
?

I can do that, but then I'll still have to address #34977 when the output of `_wrap_aggregated_output` is passed to `self.obj._constructor_expanddim(results, columns=columns)`. 

> its possible to pass columns to _wrap_aggregated_output

Are you sure? It seems to only take on argument (other than `self`)

Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py:331]: Another option would be to return `None` from `_wrap_aggregated_output`, so the return type would be `Optional[Union[Series, DataFrame]]` - would that be OK?

Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py:331]: @jreback would
```python
elif len(indexed_output):
    result = self.obj._constructor(
        indexed_output[0], index=index, name=columns[0]
    )
else:
    return None
```
be an acceptable solution?

Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py:331]: Actually,
```python
        elif not columns.empty:
            result = self.obj._constructor(
                indexed_output[0], index=index, name=columns[0]
            )
        else:
            result = self.obj._constructor_expanddim()
```
works, because
```python
pd.DataFrame(pd.DataFrame(), columns=[])
```
is allowed.

No need to modify the return types like this :)

File: pandas/core/groupby/generic.py (hunk 6 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py:362]: @WillAyd how about returning `self.obj._constructor_expanddim()` - does that seem OK?

**Comment by jreback** [comment]
this needs to go after #34998 which substantially refactors things

**Comment by MarcoGorelli** [comment]
OK, thanks for letting me know, will try rebasing locally to get this ready

**Comment by jreback** [comment]
@MarcoGorelli hmm we are going to be refactoring #34998 / delaying it, so let's see if we can get this one in. pls rebase and see if you can address above comments.

**Comment by MarcoGorelli** [comment]
@jreback have merged master

Regarding above comments, seems the only outstanding issue is handling when we pass an empty result to `self._wrap_aggregated_output`, which I've handled by returning `self.obj._constructor_expanddim()`. If that's not the right solution, any hints would be appreciated

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@TomAugspurger if you'd have a quick look.

File: doc/source/whatsnew/v1.1.0.rst [hunk 18]
```diff
@@ -1101,6 +1101,7 @@ Reshaping
 - Bug in :func:`crosstab` when inputs are two Series and have tuple names, the output will keep dummy MultiIndex as columns. (:issue:`18321`)
 - :meth:`DataFrame.pivot` can now take lists for ``index`` and ``columns`` arguments (:issue:`21425`)
 - Bug in :func:`concat` where the resulting indices are not copied when ``copy=True`` (:issue:`29879`)
+- Bug in :meth:`SeriesGroupBy.aggregate` was resulting in aggregations being overwritten when they shared the same name (:issue:`30880`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.1.0.rst:1104]: FYI: the link to this method won't render, since `SeriesGroupBy` isn't in the pands namespace.

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 16 above)
Comment (TomAugspurger) [reply pandas/tests/groupby/aggregate/test_aggregate.py:309]: Generally for this, we would just silence / filter the warning. We don't specifically care that there's a warning here.

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Thanks @MarcoGorelli!

File: doc/source/whatsnew/v1.1.0.rst (hunk 18 above)
Comment (MarcoGorelli) [reply doc/source/whatsnew/v1.1.0.rst:1104]: Sorry about that - will make sure the build the whatsnew file in the future to check

**Comment by MarcoGorelli** [comment]
This one took me a while (I opened it in January!) so thanks for bearing with me while I worked on it!
